Libraries/LibCrypto/Cipher/AES.cpp

```diff
@@ -0,0 +1,426 @@
+        if (r == 0)
+            break;
+
+        // clang-format off
+        s0 = AESTables::Decode0[(t0 >> 24)       ] ^
+             AESTables::Decode1[(t3 >> 16) & 0xff] ^
+             AESTables::Decode2[(t2 >>  8) & 0xff] ^
+             AESTables::Decode3[(t1      ) & 0xff] ^ round_keys[0];
+        s1 = AESTables::Decode0[(t1 >> 24)       ] ^
+             AESTables::Decode1[(t0 >> 16) & 0xff] ^
+             AESTables::Decode2[(t3 >>  8) & 0xff] ^
+             AESTables::Decode3[(t2      ) & 0xff] ^ round_keys[1];
+        s2 = AESTables::Decode0[(t2 >> 24)       ] ^
+             AESTables::Decode1[(t1 >> 16) & 0xff] ^
+             AESTables::Decode2[(t0 >>  8) & 0xff] ^
+             AESTables::Decode3[(t3      ) & 0xff] ^ round_keys[2];
+        s3 = AESTables::Decode0[(t3 >> 24)       ] ^
+             AESTables::Decode1[(t2 >> 16) & 0xff] ^
+             AESTables::Decode2[(t1 >>  8) & 0xff] ^
+             AESTables::Decode3[(t0      ) & 0xff] ^ round_keys[3];
+        // clang-format on
+    }
+
+    // apply the last round and put the decrypted data into out
+    // clang-format off
+    s0 = ((u32)AESTables::Decode4[(t0 >> 24)       ] << 24) ^
+         ((u32)AESTables::Decode4[(t3 >> 16) & 0xff] << 16) ^
+         ((u32)AESTables::Decode4[(t2 >>  8) & 0xff] <<  8) ^
+         ((u32)AESTables::Decode4[(t1      ) & 0xff]      ) ^ round_keys[0];
+    out.put(0, s0);
+
+    s1 = ((u32)AESTables::Decode4[(t1 >> 24)       ] << 24) ^
+         ((u32)AESTables::Decode4[(t0 >> 16) & 0xff] << 16) ^
+         ((u32)AESTables::Decode4[(t3 >>  8) & 0xff] <<  8) ^
+         ((u32)AESTables::Decode4[(t2      ) & 0xff]      ) ^ round_keys[1];
+    out.put(4, s1);
+
+    s2 = ((u32)AESTables::Decode4[(t2 >> 24)       ] << 24) ^
+         ((u32)AESTables::Decode4[(t1 >> 16) & 0xff] << 16) ^
+         ((u32)AESTables::Decode4[(t0 >>  8) & 0xff] <<  8) ^
+         ((u32)AESTables::Decode4[(t3      ) & 0xff]      ) ^ round_keys[2];
+    out.put(8, s2);
+
+    s3 = ((u32)AESTables::Decode4[(t3 >> 24)       ] << 24) ^
+         ((u32)AESTables::Decode4[(t2 >> 16) & 0xff] << 16) ^
+         ((u32)AESTables::Decode4[(t1 >>  8) & 0xff] <<  8) ^
+         ((u32)AESTables::Decode4[(t0      ) & 0xff]      ) ^ round_keys[3];
+    out.put(12, s3);
+    // clang-format on
+}
+
+void AESCipherBlock::overwrite(const ByteBuffer& buffer)
+{
+    overwrite(buffer.data(), buffer.size());
+}
+
+void AESCipherBlock::overwrite(const u8* data, size_t length)
+{
+    ASSERT(length <= m_data.size());
+    m_data.overwrite(0, data, length);
+    if (length < m_data.size()) {
+        switch (padding_mode()) {
+        case PaddingMode::Null:
+            // fill with zeros
+            __builtin_memset(m_data.data() + length, 0, m_data.size() - length);
+            break;
+        case PaddingMode::CMS:
+            // fill with the length of the padding bytes
+            __builtin_memset(m_data.data() + length, m_data.size() - length - 1, m_data.size() - length);
```

Comment:
```suggestion
            __builtin_memset(m_data.data() + length, m_data.size() - length, m_data.size() - length);
```
The AES encryption tests fail here.

This took me a while to find, but RFC 5652 section 6.3 describes the CMS padding value as "k-(lth mod k)", which should be "block_size - length" without any -1 required. This should solve the issue of non-specialized encrypted data for AES.

Note that for the PaddingMode "Bit", the value at the end should indeed be "block_size - length - 1", but that's not needed / implemented yet.

Reply:
I could swear there was a reason I did that.
perhaps I had mixed up the padding types? I'll investigate that and come back~

Reply:
It's a weirdness that RFC 5246 created, apparently TLS v1.2 uses its own padding scheme (https://tools.ietf.org/html/rfc5246#section-6.2) down in the "padding_length" section, but since it looks like CMS, I thought my CMS implementation needed tweaking.
I'll add a new "RFC5246" padding mode, I suppose.

Reply:
It's already nuts how many documents are needed to implement TLS, and then there's this kind of things... Kudos to you to make this PR happen ^^

Reply:
Yeah...the RFC-hopping was probably my least favourite part of doing this.
Issues like this one, luckily, can be found with a new perspective (I'm surprised I never ran the AES tests after starting TLS...).

It's been quite the learning experience, I'm happy I did this!